File: Python3/1419_minimum_number_of_frogs_croaking.py
# https://leetcode.cn/problems/minimum-number-of-frogs-croaking/

class Solution:
    def minNumberOfFrogs(self, croakOfFrogs):
        """
        Inputs:
            croakOfFrogs: str
        Outputs:
            res: int
        """
        res, free = 0, 0
        cnt = {ch: 0 for ch in "croak"}
        for ch in croakOfFrogs:
            cnt[ch] += 1
            if ch == "c":
                if free == 0:
                    res += 1
                else:
                    free -= 1
            elif ch == "k":
                free += 1
            else:
                for cur in "croak":
                    if cur == ch:
                        break
                    elif cnt[cur] < cnt[ch]:
                        return -1
        check = cnt["c"]
        for c in cnt.values():
            if c != check:
                return -1
        return res

File: Cpp/1419_minimum_number_of_frogs_croaking.cpp
// https://leetcode.cn/problems/minimum-number-of-frogs-croaking/

class Solution {
public:
    int minNumberOfFrogs(string croakOfFrogs) {
        int res = 0, free = 0;
        string s = "croak";
        unordered_map<char, int> cnt;
        for (char ch: s) {
            cnt[ch] = 0;
        }
        for (char ch: croakOfFrogs) {
            cnt[ch] += 1;
            if (ch == 'c') {
                if (free == 0) res += 1;
                else free -= 1;
            }
            else if (ch == 'k') {
                free += 1;
            }
            else {
                int c = cnt[ch];
                for (char cur: s) {
                    if (cur == ch) break;
                    else if (cnt[cur] < c) return -1;
                }
            }
        }
        int c = cnt['c'];
        for (char cur: s) {
            if (cnt[cur] != c) return -1;
        }
        return res;
    }
};